src/planner/Algo.h

```diff
@@ -92,6 +92,47 @@ class ProduceAllPaths final : public SingleInputNode {
     ProduceAllPaths(QueryContext* qctx, PlanNode* input)
         : SingleInputNode(qctx, Kind::kProduceAllPaths, input) {}
 };
+
+class CartesianProduct final : public SingleDependencyNode {
+public:
+    static CartesianProduct* make(QueryContext* qctx, PlanNode* input) {
+        return qctx->objPool()->add(new CartesianProduct(qctx, input));
+    }
+
+    Status addVar(std::string varName) {
+        auto checkName = [&varName](auto var) { return var->name == varName; };
+        if (std::find_if(inputVars_.begin(), inputVars_.end(), checkName) != inputVars_.end()) {
+            return Status::SemanticError("Duplicate Var: %s", varName.c_str());
+        }
+        auto* varPtr = qctx_->symTable()->getVar(varName);
+        DCHECK(varPtr != nullptr);
+        inputVars_.emplace_back(varPtr);
+        for (const auto& name : varPtr->colNames) {
+            if (std::find(allColNames_.begin(), allColNames_.end(), name) != allColNames_.end()) {
+                return Status::SemanticError(
+                    "Var : %s , exist duplicate ColName : %s", varName.c_str(), name.c_str());
+            }
+            allColNames_.emplace_back(name);
+        }
```

Comment:
Column name in `varPtr->colNames` might duplicate.

Reply:
will check it out in the for loop